app/controllers/QuestionnaireController.scala

```diff
@@ -16,110 +16,146 @@
+
+class QuestionnaireController(applicationClient: ApplicationClient) extends BaseController(applicationClient) {
 
-abstract class QuestionnaireController(applicationClient: ApplicationClient) extends BaseController(applicationClient) {
+  val QuestionnaireCompleted = "questionnaire.completed"
 
   def start = CSRSecureAppAction(StartQuestionnaireRole) { implicit request =>
     implicit user =>
-      val p = user.application.progress
       Future.successful {
-        if (!p.diversityQuestionnaire && !p.educationQuestionnaire && !p.occupationQuestionnaire) {
-          Ok(views.html.questionnaire.intro(QuestionnaireDiversityInfoForm.acceptanceForm))
-        } else {
-          Ok(views.html.questionnaire.continue())
+        (QuestionnaireCompletedRole.isAuthorized(user), QuestionnaireNotStartedRole.isAuthorized(user)) match {
+          case (true, _) => Redirect(routes.HomeController.present()).flashing(danger(QuestionnaireCompleted))
+          case (_, true) => Ok(views.html.questionnaire.intro(QuestionnaireDiversityInfoForm.acceptanceForm))
+          case _ =>   Ok(views.html.questionnaire.continue())
         }
       }
   }
 
   def firstPageView = CSRSecureAppAction(DiversityQuestionnaireRole) { implicit request =>
     implicit user =>
-      Future.successful(Ok(views.html.questionnaire.firstpage(QuestionnaireDiversityInfoForm.form)))
+      Future.successful {
+        (DiversityQuestionnaireCompletedRole.isAuthorized(user), QuestionnaireCompletedRole.isAuthorized(user)) match {
+          case (_, true) => Redirect(routes.HomeController.present()).flashing(danger(QuestionnaireCompleted))
+          case (true, _) => Redirect(routes.QuestionnaireController.start()).flashing(danger(QuestionnaireCompleted))
+          case _ => Ok(views.html.questionnaire.firstpage(QuestionnaireDiversityInfoForm.form))
+        }
+      }
   }
 
   def secondPageView = CSRSecureAppAction(EducationQuestionnaireRole) { implicit request =>
     implicit user =>
-      Future.successful(Ok(views.html.questionnaire.secondpage(QuestionnaireEducationInfoForm.form)))
+      Future.successful {
+        (EducationQuestionnaireCompletedRole.isAuthorized(user), QuestionnaireCompletedRole.isAuthorized(user)) match {
+          case (_, true) => Redirect(routes.HomeController.present()).flashing(danger(QuestionnaireCompleted))
+          case (true, _) => Redirect(routes.QuestionnaireController.start()).flashing(danger(QuestionnaireCompleted))
+          case _ => Ok(views.html.questionnaire.secondpage(QuestionnaireEducationInfoForm.form))
+        }
+      }
   }
+
   def thirdPageView = CSRSecureAppAction(OccupationQuestionnaireRole) { implicit request =>
     implicit user =>
-      Future.successful(Ok(views.html.questionnaire.thirdpage(QuestionnaireOccupationInfoForm.form)))
+      Future.successful {
+        (OccupationQuestionnaireCompletedRole.isAuthorized(user), QuestionnaireCompletedRole.isAuthorized(user)) match {
```

Comment:
This is code is pretty similar in all the steps, maybe we can have a method with a few parameters to substitute it and call it from all the steps.